File: Algorithms/BinaryTree/binarytree.py
class node:
    def __init__(self, val):
        self.key = val
        self.right = None
        self.left = None

def insert(val, root):
    if root is None:
        root = node(val)
    else:
        if val < root.key:
            root.left = insert(val, root.left)
        elif val == root.key:
            return
        else:
            root.right = insert(val, root.right)
    return root

def preorder_traversal(root, result):
    if root is not None:
        result.append(root.key)
        preorder_traversal(root.left, result)
        preorder_traversal(root.right, result)
        
with open('input.txt', 'r') as file:
    keys = [int(line) for line in file]

root = None

for key in keys:
    root = insert(key, root)

result = []

preorder_traversal(root, result)

with open('output.txt', 'w') as file:
    for key in result:
        file.write(str(key) + '\n')

File: Algorithms/BinaryTree/binarytree.cpp
#include <iostream>
#include <fstream>

struct node {
    int key;
    node* left;
    node* right;
    node(int key){
        this->key = key; left = nullptr; right = nullptr;
    }
};

using namespace std;

node * insert(node * root, int key) {
    if (root == nullptr) return new node(key);
    if (key < root->key){
        root->left = insert(root->left, key);
    }
    else if (key == root->key){
        return root;
    }
    else {
        root->right = insert(root->right, key);
    }
    return root;
}

void preorder_traversal(node * root, ofstream& out) {
    if (root != nullptr) {
        out << root->key << '\n';
        preorder_traversal(root->left, out);
        preorder_traversal(root->right, out);
    }
}

int main() {
    ifstream in("input.txt");
    ofstream out("output.txt");

    node * root = nullptr;
    while(!in.eof()) {
        int tmp;
        in >> tmp;
        root = insert(root, tmp);
    }

    preorder_traversal(root, out);
}